src/core/virtual_sites.hpp

```diff
@@ -35,54 +34,54 @@
  */
 
 #ifdef VIRTUAL_SITES
-// Recalculate position and velocity for all virtual particles
-void update_mol_vel_pos();
-// Recalc velocities for virtual particles
-void update_mol_vel();
-// Recalc positions of virtual particles
-void update_mol_pos();
-
-// The following three functions have to be provided by all implementations
-// of virtual sites
-// Update the vel/pos of the given virtual particle as defined by the real
-// particles in the same molecule
-// void update_mol_pos_particle(Particle *);
-// void update_mol_vel_particle(Particle *);
-
-// Distribute forces that have accumulated on virtual particles to the
-// associated real particles
-//void distribute_mol_force();
-
-
-// Checks, if a particle is virtual
-inline int ifParticleIsVirtual(Particle const*p){
-   if (p->p.isVirtual == 0) {
-      return 0;
-   }
-   else{
-      return 1;
-   }
-}
-
-inline int ifParticleIsVirtual(Particle  const& p){
-   return ifParticleIsVirtual(&p);
-}
-
-
-// According to what rules the virtual particles are placed and the forces and
-// torques accumulating on the virtual particles distributed back to real
-// particles, is decided by a specific implementation.
-
-// Virtual particles in center of mass of molecule
-#ifdef VIRTUAL_SITES_COM
- #include "virtual_sites_com.hpp"
-#endif
-
-// Virtual particles relative to position and orientation of a real particle
-#ifdef VIRTUAL_SITES_RELATIVE
- #include "virtual_sites_relative.hpp"
+#include <memory>
+
+/** @brief Base class for virtual sites implementations */
+class VirtualSites {
```

Comment:
The `need_*` construction seems artificial to me. These are implementation details of the VS method and should be handled internally. If you think that they should be handled generically, you can make the actual update impl private, like update_impl(), and add a update() implementation to the base class that runs the ghost updates if needed before and after the call to update_impl. If you think this is specific to a certain implementation you can also move it there. As there is currently only one real implementation this is  hard to tell for me.
  
The same applies to `back_transfer_forces_and_torques()`.